gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -350,16 +356,33 @@ public GoogleCloudStorageImpl(
     this.downscopedAccessTokenFn = downscopedAccessTokenFn;
   }
 
+  /**
+   * Constructs an instance of GoogleCloudStorageImpl. Instantiates a {@link StorageRequestFactory}
+   * from the given {@link Storage} instance.
+   *
+   * @param options {@link GoogleCloudStorageOptions} to use to initialize the object.
+   * @param storage {@link Storage} to use for I/O.
+   * @param downscopedAccessTokenFn Function that generates downscoped access token.
+   */
+  public GoogleCloudStorageImpl(
+      GoogleCloudStorageOptions options,
+      Storage storage,
+      Function<List<AccessBoundary>, String> downscopedAccessTokenFn) {
+    this(options, storage, new StorageRequestFactory(storage), downscopedAccessTokenFn);
+  }
+
   /**
    * Constructs an instance of GoogleCloudStorageImpl.
    *
    * @param options {@link GoogleCloudStorageOptions} to use to initialize the object.
    * @param storage {@link Storage} to use for I/O.
+   * @param storageRequestFactory creates storage requests that override storage api.
    * @param downscopedAccessTokenFn Function that generates downscoped access token.
    */
   public GoogleCloudStorageImpl(
       GoogleCloudStorageOptions options,
       Storage storage,
+      StorageRequestFactory storageRequestFactory,
```

Comment:
Is this just added for testing? If not, we maybe do not need it?

Reply:
If we make the change above (remove storage and add getting to get it from factory), I think we can unify the two constructor. Since there is already other an other constructor that accept a credential instead, do you think it is acceptable for now until turning all of these into a builder? I'm imagine something like this: https://softwareengineering.stackexchange.com/questions/289439/how-should-i-handle-incompatible-configurations-with-the-builder-pattern